File: Labs/Assignment4/Code/unigram_tagger.py

from collections import defaultdict
from count_freqs import Hmm
import math
import sys

def emission_probability(word, tag, emission_counts, ngram_counts):
  return emission_counts[(word, tag)] / ngram_counts[0][(tag,)]

if __name__ == "__main__":
  counts_file = open(sys.argv[1])
  sentences_file = open(sys.argv[2])
  
  hmm = Hmm()
  hmm.read_counts(counts_file)
  
  emission_counts = hmm.emission_counts
  ngram_counts = hmm.ngram_counts
  
  print ngram_counts
  sys.exit()
  
  entity_tags = ["I-PER", "I-ORG", "I-LOC", "I-MISC", "O"]
  entity_tags = hmm.all_states
  trained_words = defaultdict(int)
  infrequent_words = defaultdict(int)
  
  for word, tag in emission_counts:
    trained_words[word] += hmm.emission_counts[(word, tag)]
  
  for word in trained_words:
    if trained_words[word] < 5:
      infrequent_words[word] = 1
  
  for word in infrequent_words:
    for tag in entity_tags:
      if (word, tag) in emission_counts:
        emission_counts[("_RARE_", tag)] += 1
        del emission_counts[(word, tag)]
  
  line = sentences_file.readline()
  while line:
      word = line.strip()
      original_word = word
      if word and not(word == ""):
        if (word in infrequent_words) or not(word in trained_words):
          word = "_RARE_"
        
        current_probability = -1
        current_tag = None
        
        for tag in entity_tags:
          probability = emission_probability(word, tag, emission_counts, ngram_counts)
          if probability > current_probability:
            current_probability = probability
            current_tag = tag
            
        print original_word + " " + current_tag + " " + str(math.log(current_probability))
      else:
        print ""
      line = sentences_file.readline()
      
  

File: Labs/Assignment4/Code/unigram_tagger.rb

require 'set'
require 'pp'
require File.join(File.dirname(__FILE__), 'counter.rb')

def emission_probability(word, tag, emission_counts, ngram_counts)
  return (emission_counts[[word, tag]] || 0) / ngram_counts[0][[tag]]
end


counter = Counter.new(File.open("ner.counts"))

emission_counts = counter.emission_counts
word_counts = counter.word_counts

ngram_counts = counter.ngram_counts
entity_tags = counter.entity_tags

for word, count in word_counts do
  if count < 5
    for tag in entity_tags do
      if emission_counts.include? [word, tag] then
        emission_counts[["_RARE_", tag]] ||= 0
        emission_counts[["_RARE_", tag]] += count
      end
    end
  end
end

sentences_files = File.open("ner_dev.dat")
sentences_files.each_line do |line|
  word = line.chop
  original_word = word
  
  if word == "" then
    puts
  else
    
    if word_counts[word].nil? || word_counts[word] < 5 then
      word = "_RARE_"
    end
    
    current_probability = -1
    current_tag = nil
    
    for tag in entity_tags do
      probability = emission_probability(word, tag, emission_counts, ngram_counts)
      if probability > current_probability then
        current_probability = probability
        current_tag = tag
      end
    end
    
    
    puts original_word + " " + current_tag + " " + (Math.log(current_probability)/Math.log(2)).to_s
    
  end
  
end